### PR TITLE
Monomorphic Mapper

### PR DESCRIPTION
**Pull request by nate-thegrate**
This class looks like it _should_ give valid equality checks:

```dart
class _WidgetStateColorMapper extends _WidgetStateColorTransparent {
  const _WidgetStateColorMapper(this.map);

  final WidgetStateMap<Color> map;

  _WidgetStateMapper<Color> get _mapper => _WidgetStateMapper<Color>(map);

  @override
  Color resolve(Set<WidgetState> states) => _mapper.resolve(states);

  @override
  bool operator ==(Object other) {
    return other is _WidgetStateColorMapper && other.map == map;
  }

  @override
  int get hashCode => map.hashCode;
}
```

But I made a mistake: I should have used `other._mapper == _mapper` and `_mapper.hashCode`.

<br>

It'd be pretty nice if no copy-pasting was needed in the first place:

```dart
class _WidgetStateColorMapper extends WidgetStateMapper<Color> implements WidgetStateColor {
  const _WidgetStateColorMapper(super.map);
}
```

<br><br>

This PR makes `WidgetStateMapper` a public class, similar to `WidgetStatePropertyAll`. The new API, when used incorrectly, will `throw` with a detailed error message, rather than silently returning irrelevant values.

### CODE REVIEW

File: packages/flutter/test/widgets/widget_state_property_test.dart
Comment (MitchellGoodwin) [inline packages/flutter/test/widgets/widget_state_property_test.dart:164]: For the purpose of watching for regressions, should check if a third, different WidgetStateColor map is unequal.

Same with all of these examples.

Comment (nate-thegrate) [reply packages/flutter/test/widgets/widget_state_property_test.dart:164]: ooh, yeah that'd be good.

File: packages/flutter/lib/src/material/material_state.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/material_state.dart:309]: Also, to proactively answer a question before it's asked…

I considered matching the other types in this file, something like:

```dart
typedef MaterialStateOutlineInputBorder = WidgetStateInputBorder;
typedef MaterialStateUnderlineInputBorder = WidgetStateInputBorder;
```

<br>

The only problem with that is when someone makes a custom subclass:

```dart
class MyBorder extends MaterialStateOutlineInputBorder {} // works

class MyBorder extends WidgetStateInputBorder {} // doesn't work

class MyBorder extends OutlineInputBorder implements WidgetStateInputBorder {
  // makes no difference whether OutlineInputBorder or UnderlineInputBorder is extended
}
```

So instead we have a data-driven fix for the `.resolveWith()` constructor, and documentation that explains how to make a subclass. (Perhaps I could also update the [MaterialState migration guide](https://github.com/flutter/website/blob/main/src/content/release/breaking-changes/material-state.md) to include this info.)

File: packages/flutter/lib/src/material/material_state.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/material/material_state.dart:455]: Mentioning OutlineInputBorder in a class that is for the generic InputBorder type feels odd. Does this need to be specifically called out?

File: packages/flutter/lib/src/material/material_state.dart
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/material/material_state.dart:448]: This new class should have regression tests.

**Review by MitchellGoodwin — COMMENTED** [review]
Just reviewing the `WidgetStateInputBorder` migration before doing a separate review of the mappers. That maybe should have been it's own PR. Code LGTM. Some doc nits, and I think there needs to be tests specific to the new class

**Review by MitchellGoodwin — COMMENTED** [review]
Mapper code LGTM, no nits there.

File: packages/flutter/lib/src/material/material_state.dart
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/material_state.dart:447]: That's a great idea!

**Review by nate-thegrate — COMMENTED** [review]
@MitchellGoodwin thanks so much for the review! You're right that I should have added the mapper class in a separate PR first before doing the `WidgetStateInputBorder` stuff. Many thanks for looking over the whole thing regardless 🙏

**Review by MitchellGoodwin — APPROVED** [review]
LGTM! Thank you very much!

**Comment by nate-thegrate** [comment]
@MitchellGoodwin thanks to you for the review!

I wanted to ask a question before landing this one…

<br>

I've been thinking about potentially cherry-picking this change into the upcoming stable release:

- There might be some frustration if `.fromMap()` is introduced and then is updated 3 months later to start throwing errors more often.
- Plus, it'd be nice to not ship this out with the typo in the `WidgetStateColor.fromMap()` equality check.

Would this be worth pursuing?

**Comment by MitchellGoodwin** [comment]
> I've been thinking about potentially cherry-picking this change into the upcoming stable release:
>
> There might be some frustration if .fromMap() is introduced and then is updated 3 months later to start throwing errors more often.
Plus, it'd be nice to not ship this out with the typo in the WidgetStateColor.fromMap() equality check.
Would this be worth pursuing?

As this wouldn't be a breaking change with .fromMap() and typos in the documentation aren't critical, I don't think there's a good chance of it being cherry picked. Cherry picking is expensive on our current infrastructure team, and is usually reserved for p1/p0 issues. You can ask around in the discord for opinions from people more closely involved with releases, but I don't think it's likely to go through. Last release there were some changes that didn't make the cut which I really wanted to get through but wasn't able to 😅

**Comment by nate-thegrate** [comment]
> Cherry picking is expensive on our current infrastructure team, and is usually reserved for p1/p0 issues.

Ah, this is good to know. Thanks!

**Pull request merged**
